designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
+
+### Goals
+
+- Provide manifests that can display operator status for common needs.
+- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.
+
+### Non-Goals
+
+- TBD
+
+## Proposal
+
+This proposed implementation is to create a new plugin, `grafana`, that scaffolds the manifests which are loaded by Grafana to display dashboards for controllers' status.
+
+Basically, the manifest can be raw json file such as `controller-dashboard.json` that can be directly applied to Grafana Web UI.
+
+As there are many useful metrics, it might be better to observe in multiple dashboards rather than one.
+
+Hence, we may also provide optional dashboards that focus on different aspects.
+
+For instance: work queue, reconciliation, performance, webhook.
+
+As a result, the layout given by the plugin can be a folder of `config/grafana/`:
+
+```shell
+config/grafana
+├── controller-dashboard.json # The main dashboard raw json file for the simplest usage(copy/paste)
+├── options
+│  ├── performance.json       # CPU/Memory usaage, restful api performance
+│  ├── reconcilier.json       # Reconciliation instruments
+│  ├── webhook.jon            # Webhook Ops
+│  └── workqueue.json         # Work queue status
+└── README.md                 # Docs for the usage
+```
+
+The plugin is triggered to do the scaffolding work for the layout above when the following commands are executed:
+
+- `kubebuilder init -–plugins=grafana/alpha-v1`: when initializing a new project
+- `kubebuilder edit -–plugins=grafana/alpha-v1`: when adding features to an existing project
+
+#### Dashboard Panels
+
+The creation of the panels inside the dashboard can be referred to default [controller-runtime metrics](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/reference/metrics-reference.md).
+
+The overall dashboard should reflect:
+
+1. Latency: The duration of a request for certain controllers/operators, which includes both successful ops and errors ops.
+2. Traffic: The measure of how busy the controllers/operators can be.
+3. Errors: Error rates, or the operations/requests that takes unexpectedly long time.
+4. Saturation: The measure of the possibility of overflows such as deep work queue depth, concurrencies of reconciliations, large cpu/memory usage.
+
+Also, the dashboard should provide filters and selectors to focus on certain objects. This can be implemented by utilizing [Grafana Query Variables](https://grafana.com/docs/grafana/latest/datasources/prometheus/#query-variable).
```

Comment:
These queries will be specified in JSON files, or the user needs to filter them manually in a dashboard?